### PR TITLE
posix/unix: add unix sockets polling

### PR DESCRIPTION
**Pull request by ziemleszcz**
JIRA: DTR-240

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/85).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: posix/unix.c [hunk 1]
```diff
@@ -947,6 +947,46 @@ int unix_close(unsigned socket)
...
+	else {
+		if (events & (POLLIN | POLLRDNORM | POLLRDBAND)) {
+			proc_lockSet(&s->lock);
+			if (_cbuffer_avail(&s->buffer) > 0)
+				err |= events & (POLLIN | POLLRDNORM | POLLRDBAND);
+			proc_lockClear(&s->lock);
+		}
+
+		if (events & (POLLOUT | POLLRDNORM | POLLRDBAND)) {
+			if ((conn = s->connect) != NULL) {
```
Comment (nalajcie) [inline posix/unix.c:967]: i'm wondering if we could have a race condition here (memory pointed by `conn` could be freed after this line by doing `unixsock_put` (eg. by `unixsock_close`) on the connected socket. I think that:
 - we should ensure every unixsock_put takes lock to closing socket or is called under lock
 - to modify `s->connect` in `unixsock_put` we should take `s->lock` (and every access to s->connect should be also done under the lock?)
 - (I don't know if we won't introduce some way to deadlock sockets closing, but it's done under the global lock so maybe we're safe - I don't know the usual code paths here)

Edit: maybe better would be to use `unixsock_get(s->connect_fd)` to ensure the connected socket won't be deleted while we're operating on it? (seems like a cleaner solution to me)

Comment (ziemleszcz) [reply posix/unix.c:967]: Yeah, good catch. Unfortunately this code is probably full of race conditions. I don't wont to dive into this right now. In this particular case your "unixsock_get" suggestion should fix it. Another option is to acquire "s->lock" in "unixsock_put" before destroying the socket - this should also fix the same problem in "send" function.

Comment (nalajcie) [reply posix/unix.c:967]: Yeah, IMHO we should always use `unixsock_get` when we're operating on a peer socket, but apart from the poll it's out of scope of this PR (maybe we should add a task not to forget it?). Socket deletion should also be reviewed. I was thinking about switching field `connected` to `connected_fd`, but this is not enough to fix deletion use case (and would result in strange behavior if somebody would close the connected socket and open new one reusing the FD :)).

I'll go for using `unixsock_get` here and at least adding FIXME in places we suspect race condition may occur.

Comment (nalajcie) [reply posix/unix.c]: IMHO there still could be a race condition (`s->connect` could be freed between  `s->connect != NULL` access and  && `conn = unixsock_get(s->connect->id)` - maybe add FIXME there? But now the window for race condition is much smaller :)

**Review by nalajcie — APPROVED** [review]

File: posix/unix.c [hunk 2]
```diff
@@ -218,13 +218,30 @@ static unixsock_t *unixsock_get(unsigned id)
...
+{
+	unixsock_t *r;
+
+	proc_lockSet(&unix_common.lock);
+	if (s->connect) {
+		r = s->connect;
+		r->refs++;
+	}
+	else
+		r = NULL;
```
Comment (agkaminski) [inline posix/unix.c:231]: Coding style - missing bracket for `else`

File: posix/unix.c [hunk 3]
```diff
@@ -218,13 +218,30 @@ static unixsock_t *unixsock_get(unsigned id)
 }
 
 
+static unixsock_t *unixsock_get_connected(unixsock_t *s)
+{
+	unixsock_t *r;
+
+	proc_lockSet(&unix_common.lock);
+	if (s->connect) {
```
Comment (agkaminski) [inline posix/unix.c:226]: Coding style
```suggestion
	if (s->connect != NULL) {
```

File: posix/unix.c [hunk 4]
```diff
@@ -218,13 +218,30 @@ static unixsock_t *unixsock_get(unsigned id)
...
+		r = s->connect;
```
Comment (agkaminski) [inline posix/unix.c:227]: We assign `r = s->connect` in both cases, this can be simplified

**Review by agkaminski — APPROVED** [review]

**Pull request merged**